### PR TITLE
🌱   refactor: standardize machine filter functions and improve testing

### LINKED ISSUE #2062: util package: Make functions that fetch/filter many Machines consistent with one another
The `util` package has functions that fetch or filter many machines. Some functions use the `clusterv1.MachineList` type, others use `[]*clusterv1.Machine`. Let's standardize on one.

(Issue came out this discussion: https://github.com/kubernetes-sigs/cluster-api/pull/2037#discussion_r366055528)

/priority important/long-term

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What this PR does / why we need it**:  The util, controllers, kcp internal packages had overlapping functions that fetch or filter many machines. Some functions use the clusterv1.MachineList type, others use []*clusterv1.Machine, others use collections.Machines. This PR standardizes the use of filter functions in util/collections across the codebase.

(Issue came out this discussion: #2037 (comment))

In addition, I found a few "bugs" in tests of those functions - tests that weren't actually testing what they said they were testing so I fixed those.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2062

### CODE REVIEW

File: controllers/machine_controller_test.go
Comment (CecileRobertMichon) [inline controllers/machine_controller_test.go:1306]: this tests was not fully correct as the machines were being excluded from the filter by not having the right cluster name label and/or namespace. Added all the required object metadata so it's actually checking that the described missing piece (eg. No nodeRef) is what causes the error.

File: controlplane/kubeadm/controllers/status_test.go
Comment (CecileRobertMichon) [inline controlplane/kubeadm/controllers/status_test.go:49]: this one got me debugging the test for a while 😬 

the test was actually a no-op because of https://github.com/kubernetes-sigs/cluster-api/pull/4207/files#diff-a6639fd91ffb9cff50ef1f3af76ec2371aa5c41e0d28b98b651cb9193eeaab62L52

Since the fake management cluster `Management ` was nil, it wasn't actually listing machines and filtering them, it was just returning ALL machines that were passed into managementCluster.Machines. When I changed it to actually using the filter, it was failing because the Machines weren't "owned" by KCP since the KCP TypeMeta was missing here but not in the owner ref, it was comparing empty string to the ref's Kind, Group, and Version.

File: util/util.go
Comment (CecileRobertMichon) [inline util/util.go:112]: there's some overlap with #4005 here, I don't mind rebasing on top of it once it merges.

This is ready for review in the meantime though.

/hold

**Comment by CecileRobertMichon** [comment]
```
sigs.k8s.io/cluster-api/util
  Incompatible changes:
  - HasPausedAnnotation: removed
  - ImageTagIsValid: removed
  - IsPaused: removed
  - MachineListFormatDeprecationMessage: removed
  - ModifyImageRepository: removed
  - ModifyImageTag: removed
  - PointsTo: removed
  - SemverToOCIImageTag: removed
  - WatchOnClusterPaused: removed
  ```
  
  All of should be removed in https://github.com/kubernetes-sigs/cluster-api/pull/4078 already

File: controllers/cluster_controller.go
Comment (vincepri) [inline controllers/cluster_controller.go:405]:
```suggestion
        machineColleciton := collections.FromMachineList(&machines)
	controlPlaneMachines := machineCollection.Filter(collections.ControlPlaneMachines(cluster.Name))
	workerMachines := machineCollection.Difference(controlPlaneMachines)
	descendants.workerMachines = collections.ToMachineList(workerMachines)
```

File: util/collections/machine_filters.go
Comment (vincepri) [inline util/collections/machine_filters.go:221]: Do we want to use a client within this package?

Comment (CecileRobertMichon) [reply util/collections/machine_filters.go:221]: it was already being used in the same func in the util package https://github.com/kubernetes-sigs/cluster-api/pull/4207/files#diff-a61d5e39f45fb5ed2dc1ef71fa708c8db480e9a396adb8a5d9d17b8e6ba1e186L129, this moves it to a more specific sub package. Would there be a better place to put it if we don't want any client usage in util/?

File: controllers/machine_controller.go
Comment (MarcelMue) [inline controllers/machine_controller.go:469]: Would it make sense to simplify this to an `if / else` statement? Would be a bit easier to read for me at least with only 2 outcomes.

File: util/collections/machine_collection.go
Comment (MarcelMue) [inline util/collections/machine_collection.go:62]:
```suggestion
// ToMachineList creates a MachineList from the given Machines.
```

File: util/collections/machine_filters.go
Comment (fabriziopandini) [reply util/collections/machine_filters.go:221]: What about moving to a helper.go file under utils/collections (It is a little bit confusing having this func into machine_filters)

**Comment by CecileRobertMichon** [comment]
```
sigs.k8s.io/cluster-api/util
  Incompatible changes:
  - ImageTagIsValid: removed
  - MachineListFormatDeprecationMessage: removed
  - ModifyImageRepository: removed
  - ModifyImageTag: removed
``` 

All 4 are deprecated in alpha3: https://github.com/kubernetes-sigs/cluster-api/blob/release-0.3/util/util.go

File: controlplane/kubeadm/controllers/controller.go
Comment (CecileRobertMichon) [reply controlplane/kubeadm/controllers/controller.go:273]: let me look into this.

File: util/util.go
Comment (CecileRobertMichon) [reply util/util.go:107]: rebased on top of #4078

/hold cancel

File: controlplane/kubeadm/controllers/controller.go
Comment (CecileRobertMichon) [reply controlplane/kubeadm/controllers/controller.go]: changed it back to use r.managementClusterUncached and instead modified the implementation, PTAL

File: util/collections/machine_filters.go
Comment (CecileRobertMichon) [reply util/collections/machine_filters.go]: done

**Comment by CecileRobertMichon** [comment]
@MarcelMue @fabriziopandini thanks for the reviews. Now that #4078 has merged I've rebased this PR and addressed your comments, PTAL.

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Pull request merged**
